pkg/reconciler/reconciler_test.go

```diff
@@ -1459,6 +1557,39 @@ func verifyHooksCalled(ctx context.Context, r *Reconciler, req reconcile.Request
 	})
 }
 
+type annotatingPostRenderer struct {
+}
+
+func withAnnotatingPostRenderer(_ context.Context, _ PostRendererContext) (postrender.PostRenderer, error) {
+	return &annotatingPostRenderer{}, nil
+}
+
+func (a annotatingPostRenderer) Run(renderedManifests *bytes.Buffer) (modifiedManifests *bytes.Buffer, err error) {
+	// Add foo=bar annotation to the rendered manifests.
+	objs := manifestToObjects(renderedManifests.String())
+	var manifests = make([]string, len(objs))
+	for i, obj := range objs {
+		var annotations = obj.GetAnnotations()
+		if annotations == nil {
+			annotations = map[string]string{}
+		}
+		annotations["foo"] = "bar"
+		obj.SetAnnotations(annotations)
+		manifest, err := yaml.Marshal(obj)
```

Comment:
One of the ways we could do this is using the [controller-runtime interceptor](https://github.com/kubernetes-sigs/controller-runtime/tree/main/pkg/client/interceptor), intercept the create/update calls, and handle structured and unstructured objects to append the annotations before creating the objects. 

Though in stackrox's desired usecase ([apply user-provided overlay patches](https://github.com/stackrox/k8s-overlay-patch)), this could only be done, AFAIK, using de & reserialization.

Reply:
Oh, right, it's for arbitrary modifications, not just annotations. I updated my comment above.

However given that this library [already does](https://github.com/stackrox/helm-operator/blob/08cc94506b32b820d75179190dd1f3fc1c287bd1/pkg/client/postrenderer.go#L115) a deserialize+reserialize in the `ownerPostRenderer` I think it might be better to refactor the code there to allow adding custom visitors that could walk already parsed objects, rather than custom postrenderers that would repeat the same work?

Reply:
I'm not sure, because the user might want to remove parts of the manifest altogether, or add new resources to it, which would not be possible with a `walk`. 
I ran some benchmark tests, and the impact seems negligible. Deserializing then re-serializing an object from-to a yaml manifest takes roughly 0.0003 sec. I ran a test with 1000 deployments in 1 manifest (extremely large manifest), and it takes 0.27 sec.

Reply:
Good point about visitor not being powerful enough for some use cases!

However I still think this change would be exposing functionality at the wrong level of abstraction.
It looks like the `PostRenderer` [hook exposed in the helm library](https://github.com/helm/helm/blob/f099c2e883f43f853cdffd974ad2b8f1792cee1c/pkg/postrender/postrender.go#L24) works on the level of `bytes.Buffer` objects because this is a natural fit [for exposing it](https://github.com/helm/helm/blob/f099c2e883f43f853cdffd974ad2b8f1792cee1c/cmd/helm/flags.go#L117) from the `helm` CLI - since that works on the level of command pipelines.

While this works, and the overhead seems manageable (thanks for checking this!), I think it forces the API user to do unnecessary work, so is not elegant.

What do you think about exposing a [`kube.ResourceList`](https://github.com/helm/helm/blob/f099c2e883f43f853cdffd974ad2b8f1792cee1c/pkg/kube/resource.go#L22) instead? (That one has `Append` and `Filter`, for example).